### ISSUE TITLE
Related resources not fetched

### ISSUE THREAD

**Issue by ashishgodani** [katharsis-project/katharsis-core#204]
I am trying my hands-on katharsis and I and a little confused with the way a composed objects are fetched. A sample scenarios goes like: I would like to fetch project details along with the task in one go so 'data' should have the value for the project ```{"id": "10","name": "test project"}```. Expected sample response will look like (using spring-boot-simple-example): 
```json
{
"data": {
"type": "tasks",
"id": "10",
"attributes": {
"my-name": "Some task"
},
"relationships": {
"project": {
"links": {
"self": "http://localhost:8080/tasks/10/relationships/project",
"related": "http://localhost:8080/tasks/10/project"
},
"data": {"id": "10","name": "test project"}
}
},
"links": {
"self": "http://localhost:8080/tasks/10"
}
},
"included": []
}
```
How do I achieve this ? I tried the samples available in the repo but none of them seems to cater to this requirement. I was hoping that JsonApiIncludeByDefault would do this but it doesnt fetch the complex type attributes or I am doing something wrong:
```java
    public class Task {

    @JsonApiId
    private Long id;

    @JsonProperty("my-name")
    private String name;

    @JsonApiToOne
    @JsonApiIncludeByDefault
    private Project project;

```

Appreciate any help in this regard - prefer an example if possible.
Thanks!

**Comment by meshuga** [comment]
well, when you are making a `GET` request, all of the attributes and relationships are fetched from a `ResourceRepository`. Is your repository return the relationships?
If you add `JsonApiIncludeByDefault` annotation, the relationship if fetched (if not present using `ResourceRepository`) from a `RelationshipRepository` and additionally is in `include` attribute in the response. Have you implemented `RelationshipRepository` properly?

**Comment by ashishgodani** [comment]
Thanks for the quick response. Thats what my thinking also was but for a strange reason I am not able to make it work. 
Code for Task is already listed above.
RelationshipRepo:

    public class TaskToProjectRepository implements RelationshipRepository<Task, Long, Project, Long> {

    private TaskRepository taskRepository;
    private ProjectRepository projectRepository;

    public TaskToProjectRepository() {
      //default constructor
    }

    @Autowired
    public TaskToProjectRepository(TaskRepository TaskRepository, ProjectRepository projectRepository) {
      this.projectRepository = projectRepository;
      this.taskRepository = taskRepository;
    }

    @Override
    public void setRelation(Task task, Long projectId, String fieldName) {

    }

    @Override
    public void setRelations(Task task, Iterable<Long> projectId, String fieldName) {

    }

    @Override
    public void addRelations(Task source, Iterable<Long> targetIds, String fieldName) {
    }

    @Override
    public void removeRelations(Task source, Iterable<Long> targetIds, String fieldName) {
    }

    @Override
    public Project findOneTarget(Long sourceId, String fieldName, QueryParams requestParams) {
        return projectRepository.findOne(sourceId, requestParams);
    }

    @Override
    public Iterable<Project> findManyTargets(Long sourceId, String fieldName, QueryParams requestParams) {
        return projectRepository.findAll(requestParams);
    }
    }

and this is the project resource:

    public class Project {

    @JsonApiId
    private Long id;

    private String name;

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public Project(Long id, String name) {
        this.id = id;
        this.name = name;
    }
    }

**Comment by ashishgodani** [comment]
Hi! I have attached the sample code for my springboot-redis(or any key-value store) scenario.

Just to give the context - basically my requirement is to have this working with a key-value (say redis) store. I have namespaces in redis like ```project:1``` or ````task:1````where the data is stored in JSON format. These JSON objects have like one-to-one relationship - for example: each task will have only one project but they dont know each other at all. 
Task and Project dont know each other at all - no relationship!
Its in the code that we try to compose these relationships by fetching them individually with the same id - like fetch task with id=1 then fetch project with id=1 and finally do a setProject() on task to set the relationship.

I would like to realize this using katharsis in spring-boot, but I am not able to :( 
It looks like I am missing something really simple to populate the relationship or I didnt understand the concept at all. Any pointers will be really helpful. Thanks!

[SpringBootSampleCode.zip](https://github.com/katharsis-project/katharsis-core/files/83047/SpringBootSampleCode.zip)

**Comment by meshuga** [comment]
Just initialize a relationship between two resources and it should work.

```java
@Component
public class InitilizeRepo {

     // ...

    @PostConstruct
    public void initialize() {
        // initialize resources
        taskOne.setProject(projectOne);
        taskRepository.save(taskOne); // just pretend we save to real DB
        taskTwo.setProject(projectTwo);
        taskRepository.save(taskTwo); // just pretend we save to real DB
    }
}
```

You could also add `JsonApiIncludeByDefault` annotation to ease the way the request is made:
```java
@JsonApiResource(type = "tasks")
public class Task {

    // ...

    @JsonApiToOne
    @JsonApiIncludeByDefault
    private Project project;
}
```

**Comment by ashishgodani** [comment]
Thanks again for the response!

It still doesnt work if the full composition is not persisted in the database. Basically I commented ```repo.put(entity.getId(), entity);``` in ```TaskRepository.save()``` method in the repo as I dont want to persist the full composition into the DB.

and I modified ```initialize()``` to ```task.setProject(project)``` then call save to initialise the relationship between the two resources:
```java
public class InitilizeRepo {

    // ...

    @PostConstruct
    public void initialize() {

        Task taskThree = new Task(3L, "Task Three");
        Project projectThree = new Project(3L, "Project Three");
        taskThree.setProject(projectThree);
        taskRepository.save(taskThree);
    }
}
```
[SpringBootSampleCode_update1.zip](https://github.com/katharsis-project/katharsis-core/files/83434/SpringBootSampleCode_update1.zip)

**Comment by ashishgodani** [comment]
Hi Patryk! 
I did a few more experiments to make the relationship fetch work out of the box by just annotating the domain object but it just doesnt seem to work. Infact `findOneTarget()` method (annotated as `@JsonApiFindOneTarget`) is not even called once when I fetch a Task by calling GET - http://localhost:8080/tasks/1 
however this works **if and only if** the related object, ```Project``` in this scenario, is persisted and fetched along with the task itself. This being the case it expects the domain object to be persisted as a self contain one. This way it cannot be composed from different sources - like may be another service call to get the relationship.

I still think I am missing something in the implementation which is causing this behaviour but I am not able to figure out myself and I dont see anyone facing this issue earlier to refer-to.

So I would like to know if I initialized the relationship correctly as suggested by you ? If not - whats the right way to do it & practically how do we initialize the resource-relationship because it may not be possible always to save a dummy value to the db.
and if yes - is this how it is supposed to work where the `Task` wont fetch the related `Projects` in this scenario  ?
So basically - how do I make it work ? 

Again, thanks for the timely support so far!
[SpringBootSampleCode_update2.zip](https://github.com/katharsis-project/katharsis-core/files/85850/SpringBootSampleCode_update2.zip)

**Comment by ashishgodani** [comment]
Well, this was the case when no ORM tool was used to fetch the related entities however using an ORM framework would take care of this automatically.

Basically this was a misunderstanding as to what the framework was meant to do - resolved/clarified now!
